### PR TITLE
Site Editor: Improve loading experience (v3)

### PR DESCRIPTION
**Pull request by tyxla**
## What?

This experiments with loading the site editor behind the scenes and displaying it once it's finished loading. This time, we're doing it by checking whether we have any resolving selectors belonging to the core data store.

This PR is an alternative to #42525 and #47612, but this time, we don't utilize [Suspense](https://react.dev/reference/react/Suspense).

## Why?

We're aiming to find a way to improve the editor loading experience.

As @mtias [says](https://github.com/WordPress/gutenberg/issues/35503#issuecomment-973182125):

> The end goal should be that things load like a "screenshot"; all fully loaded when you start interacting.

See #35503 for more details and motivation.

## How?

We're building on what we've learned from #42525 and #47612. We're reusing the existing loading screen. 

### Demo - zoomed-out mode loading

https://user-images.githubusercontent.com/8436925/235678505-acbfb500-aea3-4d23-b0a7-796ce65d6d86.mov

### Demo - editor canvas loading

https://user-images.githubusercontent.com/8436925/235678660-36a0d5f3-5a4a-43b5-a5c9-f4019dffebb7.mov

## Testing Instructions
* Load the site editor and observe the loading experience. 
* Observe that at the time the fallback placeholder has been hidden, the editor has fully loaded.
* Test various use cases, like loading the zoomed-out site editor mode and the editor canvas mode.

### CODE REVIEW

**Comment by youknowriad** [comment]
I like the simplicity of this and I do see us potentially using it for the initial loading in the site editor but unfortunately, it has one major drawback. We can't know which block editor instance triggered the requests, so we can't integrate it into patterns loading / templates loading... and that's also an important use-case IMO.

**Comment by tyxla** [comment]
> We can't know which block editor instance triggered the requests, so we can't integrate it into patterns loading / templates loading... and that's also an important use-case IMO.

Could you please provide more details about the flow you're referring to? From my understanding, this loading experience should be usable for editing templates as well, but I assume you mean something else when you say "templates loading" and "patterns loading". Ideally, step-by-step instructions would be very helpful to ensure we're talking about the same thing.

**Comment by youknowriad** [comment]
@tyxla There are multiple possibilities:

 - In the inserter, in the patterns tab. When you open a category like "Posts" or "Headers", each pattern preview should have a loader of its own. So for instance if there's a pattern that uses a query, the pattern needs to show a loader until all its blocks have loaded properly.

 - Maybe more visible, but we have an API to define "starter templates" or "starter full page patterns". For instance, you can go to the site editor, click "templates" in the sidebar, click the "plus" button to add a new template and choose like "tag", you will be provided with a mosaic of templates (probably only one template for now + a start blank option as most themes didn't start using the APIs yet), each template should have a loader and only render when its content is ready (query blocks, navigation)...

Does that help?

**Comment by tyxla** [comment]
@youknowriad thanks, it does help! 

The intention of this PR isn't to cover all those cases, though. 

We'd be happy to explore solutions for improving their loading experiences separately.

**Comment by tyxla** [comment]
All tests are now green and I think this is ready for a review at this point. I believe this is far from perfect, but to a degree fulfills our initial intent. Any feedback is welcome!

**Comment by jasmussen** [comment]
Thanks for working on this, this is an important one. I know that @jameskoster has a design for a progressbar that eventually we want to build and apply, that can possibly be done in a few places, but can probably also be separate as the spinner can work in the mean time. 

Quick GIF for conversation:

spinner

A few things come to mind. The initial loading does succeed in hiding some of the awkward HTML loading of the interface getting built. But it's also quickly over, making way for yet another spinner in the preview, and for the navigation on the left to still load. I wonder if we shouldn't also preload _at least_ the left navigation, before removing the main overlay, and the preview as well. 

The preview can still retain its own spinner, that is still useful for when you are in the site editor and loading a different template. But for the initial load, you should ideally only ever see a single spinner/progress-bar, and then it's loaded. Even if it takes a bit longer.

Secondly, it should ideally have the same background color as the site editor itself. That's an easy color for now, but be aware that when some theming lands, we want it to retain the same themed background color of the site editor, not just be locked to a dark gray. A solution, potentially (I will defer to Jay who has [some previous mockups on this](https://github.com/WordPress/gutenberg/issues/35503#issuecomment-1022401146)) is to not treat this as an overlay, but just as keeping everything at opacity 0 until loaded, at which point it can fade in quickly. That would keep the background gray for the post editor, dark gray for the site editor. Or more accurately, not have a background color of its own, and just inherit what's offered by each section that loads.

**Comment by jameskoster** [comment]
Looking at the gif above it seems sub-optimal that the left sidebar loads piecemeal. This is essentially the issue we're trying to fix in the frame. 

> you should ideally only ever see a single spinner/progress-bar, and then it's loaded

Generally I agree, but this does prohibit any interaction with the sidebar until the preview is loaded, which could be a source of frustration?

All the original mockups around this were based only on the full-screen edit view. It might be worth revisiting that now that we have the zoomed-out site view as well:

1. Mock up separate loaders for sidebar / frame
2. Mock up single loader for the entire app

File: packages/edit-site/src/components/editor/index.js
Comment (kevin940726) [inline packages/edit-site/src/components/editor/index.js:178]: Curious to know what this arbitrary timeout is for?

File: packages/edit-site/src/components/editor/index.js
Comment (kevin940726) [inline packages/edit-site/src/components/editor/index.js]: We can just move the cleanup function inside the if statement where we set the timeout, so that we don't need the second if statement.

**Comment by jameskoster** [comment]
Here are a couple of quick mockups to explore the combined versus separate loaders for the sidebar / frame:

### Combined

https://user-images.githubusercontent.com/846565/235910000-129d1630-4c81-41a0-82a3-6e3a960528ef.mp4

### Separate

https://user-images.githubusercontent.com/846565/235910039-41683a70-ca86-4d2a-ba0b-8814fd17a934.mp4

As mentioned above, the benefit with this approach is that you can begin interacting with the sidebar before the frame is loaded, which could take some time.

It also seems possible that the frame will need an independent loading state anyway, if we're going to have interactions such as hovering an item in the sidebar and viewing a corresponding preview in the frame.

What do y'all think @WordPress/gutenberg-design

**Comment by tyxla** [comment]
Thanks for taking a look, @jasmussen!

> I know that @jameskoster has a design for a progressbar that eventually we want to build and apply, that can possibly be done in a few places, but can probably also be separate as the spinner can work in the mean time.

FWIW this doesn't introduce any loading UI, it only reuses what was already being used. I definitely agree that what we display there can be improved in favor of a better progress bar. 

> A few things come to mind. The initial loading does succeed in hiding some of the awkward HTML loading of the interface getting built. But it's also quickly over, making way for yet another spinner in the preview, and for the navigation on the left to still load. I wonder if we shouldn't also preload at least the left navigation, before removing the main overlay, and the preview as well.

Noting that this PR doesn't really change the initial loading outside the frame. It was already there and while I agree with you that it can be improved too, I think it can and should be explored in a separate PR. 

> The preview can still retain its own spinner, that is still useful for when you are in the site editor and loading a different template. But for the initial load, you should ideally only ever see a single spinner/progress-bar, and then it's loaded. Even if it takes a bit longer.

Yep, this PR mostly focuses on the loading experience inside the preview, where previously each block could have a its own spinner, and now we have just one spinner. 

Are you saying that for the zoomed-out mode, you'd like a single spinner and after that everything, including the preview, has loaded? The preview can take a while more to load sometimes, and as a user I can see some value in having the sidebar loaded first, and then the preview to have its own spinner.

> Secondly, it should ideally have the same background color as the site editor itself. That's an easy color for now, but be aware that when some theming lands, we want it to retain the same themed background color of the site editor, not just be locked to a dark gray. A solution, potentially (I will defer to Jay who has some previous mockups on this) is to not treat this as an overlay, but just as keeping everything at opacity 0 until loaded, at which point it can fade in quickly. That would keep the background gray for the post editor, dark gray for the site editor. Or more accurately, not have a background color of its own, and just inherit what's offered by each section that loads.

Sounds like a plan to me, we could explore this. However, I'd like to note once more that this doesn't build any new UI, and the loading spinner with the white background was already there. I'd expect any changes to that to be explored in their own PR.

**Comment by tyxla** [comment]
Thank you for looking, @jameskoster!

> Looking at the gif above it seems sub-optimal that the left sidebar loads piecemeal. This is essentially the issue we're trying to fix in the frame.

As I was saying to Joen above, the preview can take a while more to load sometimes, and as a user I can see some value in having the sidebar loaded first, and then the preview to have its own spinner. We could, of course, keep a single spinner on top of everything, but do we want to perhaps show something to the user in the meantime? The zoomed-out sidebar is a good compromise IMHO because it doesn't really cause any layout shifts, it only adds a few menu items during loading.

> Generally I agree, but this does prohibit any interaction with the sidebar until the preview is loaded, which could be a source of frustration?

I think I agree with this, given that the entire loading experience could take 20-30 seconds or even more on slower machines / connections.

> All the original mockups around this were based only on the full-screen edit view. It might be worth revisiting that now that we have the zoomed-out site view as well:

Agreed. I still think that what we're doing here is an improvement for the full-screen edit view, and for the preview in the zoomed-out view and I'd love to have it land. We can later explore other opportunities to improve the loading experience in the context of zoomed-out view. I'm happy to aid with some of those experiments.

> As mentioned above, the benefit with this approach is that you can being interacting with the sidebar before the frame is loaded, which could take some time. It also seems possible that the frame will need an independent loading state anyway, if we're going to have interactions such as hovering an item in the sidebar and viewing a corresponding preview in the frame.

**Comment by tyxla** [comment]
> Here are a couple of quick mockups to explore the combined versus separate loaders for the sidebar / frame:

Thanks for those, @jameskoster! I'd love to hear some more design feedback before we move forward with one of those.

I'd like to note that this doesn't really change the top-level loader and only affects the one inside the preview, or if loading the editor in full-screen mode. I'm happy to take suggestions and consider addressing them separately in another PR. It doesn't have to be that we're improving the entire experience in just one step after all.

File: packages/edit-site/src/components/editor/index.js
Comment (tyxla) [reply packages/edit-site/src/components/editor/index.js:178]: There are brief moments where there aren't any resolving selectors, but a few milliseconds later, there are resolving selectors present. This arbitrary timeout is there to ensure that we're marking the editor as loaded only if there haven't been any resolving selectors for an entire second.

File: packages/edit-site/src/components/editor/index.js
Comment (tyxla) [reply packages/edit-site/src/components/editor/index.js]: Agreed, moved in 8f5790ea5e0cca3f8e1d76271c0039ed6d8a57e2

**Comment by jasmussen** [comment]
> What do y'all think https://github.com/orgs/WordPress/teams/gutenberg-design

> Thanks for those, @jameskoster! I'd love to hear some more design feedback before we move forward with one of those.

I like both your designs, Jay.

But I also think they are good followups that should not block this PR. The PR as is, is a big improvement over trunk.

I'd personally like to see the white screen flash addressed as a followup. I genuinely thought that was a part of this PR, my apologies, it is not. So in fact, simply just landing this PR that adds a spinner to the main frame, that feels like a great first step, and then we can extract Jay's mockups and update #35503 with them. What do you think?

**Comment by mtias** [comment]
One comment on the "combined" mockup: it still seems like the frame and sidebar can be painted in the loading state.

**Comment by jameskoster** [comment]
The spinner placement seems a little odd to me, but it could work.


https://user-images.githubusercontent.com/846565/235982723-ad0be2df-b493-4e94-b5ce-26eed0a4bc6f.mp4

File: packages/edit-site/src/components/editor/index.js
Comment (kevin940726) [reply packages/edit-site/src/components/editor/index.js:178]: This reads a bit like an anti-pattern to me and a bad UX for users. Now every user has to wait at least a second to load the editor, no matter how fast their device is. Could there also be cases when the user's device is so slow that the 1-second delay isn't sufficient to run all the remaining tasks?

Could we figure out the correct timing to subscribe to the load event under the hood? This might be a bit more difficult to do but I think it'll be worth it 😆. At the very least, let's add some comments to explain the timeout.

Comment (talldan) [reply packages/edit-site/src/components/editor/index.js:178]: In terms of blocks, it seems like it's difficult to know what's going to load next. The next resource/block to load is revealed only once the previous resource/block has been loaded, parsed and rendered. My assumption is that's what this timeout is for.

The two ideas I had:
- Be able to parse block content to determine what resources will load next. This would require a consistent way to declare resources for blocks. Only hide the loading spinner when there are no more known resources and all resolution has finished.
- Somehow wait until all known inner blocks have rendered (and assume those blocks will have made their next requests and aren't waiting to do so).

**Comment by tyxla** [comment]
@mtias @jameskoster @jasmussen From the design perspective, how do you folks feel about landing the loading experience improvement for the full-screen editor canvas as-is? 

Then we could separately work on adding a similar loading experience for the zoomed-out view, following the same strategy and reusing some of the added functionality of this PR. 

I do appreciate the improvements suggested in the latest mockups, I just think that they can be done independently and separately and both approaches for improving the loading experience can be beneficial since they're utilized in different ways and in different contexts.

Thoughts?

File: packages/edit-site/src/components/editor/index.js
Comment (tyxla) [reply packages/edit-site/src/components/editor/index.js:178]: > This reads a bit like an anti-pattern to me and a bad UX for users. Now every user has to wait at least a second to load the editor, no matter how fast their device is. Could there also be cases when the user's device is so slow that the 1-second delay isn't sufficient to run all the remaining tasks?

I don't claim for the 1s to be perfect. I claim it to be much better than what we had before, and much better than if we didn't have it all (in which case we'd be seeing bad flickering).

Yes, 1s might be insufficient for very slow devices, but we're making a compromise here and I'm happy to experiment with timing separately. Regardless, I think this is much better than what we had before, so it's worth going with it for the time being while experimenting with better timing.

> Could we figure out the correct timing to subscribe to the load event under the hood? This might be a bit more difficult to do but I think it'll be worth it 😆. At the very least, let's add some comments to explain the timeout.

I'm open to suggestions. From my understanding, there isn't a good way to predict or subscribe to that, nor to figure out the perfect timing without a lot of experimentation and speculation. @talldan has explained that well in the above comment. 

In the meantime, I've added a comment to explain the arbitrary timeout: c7fed44b660af7f940461ad50cc9f3886a7c9ded

> The two ideas I had:

Both of those seem pretty complex to me, and I'm not sure it's worth investing in an additional complex parsing mechanism or a complex loading tracking mechanism, given that we're aiming to start async loading blocks at some point in the near future (with the upcoming work on ES modules and import maps adoption in the WordPress core).

**Comment by mtias** [comment]
@tyxla from a design flow perspective, I'm completely fine with this as the first step as it fixes the biggest problem and makes the experience a lot more serene. Thanks!

**Review by mtias — APPROVED** [review]

**Comment by jasmussen** [comment]
🚀

**Comment by tyxla** [comment]
Thanks, everyone for the excellent design feedback!

Any other feedback on the code before we go ahead with this one?

File: packages/edit-site/src/components/editor/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/editor/index.js:186]: Potentially this logic to compute the isLoading const could be extracted to a custom hook `useSiteEditorLoading` or something.

File: packages/data/src/redux-store/metadata/selectors.js
Comment (youknowriad) [inline packages/data/src/redux-store/metadata/selectors.js:142]: Depending how confident we are with this selector (it seems fine to me), we could consider making it private to start with.

**Review by youknowriad — COMMENTED** [review]
Code wise this looks good to me.

File: packages/edit-site/src/components/editor/index.js
Comment (tyxla) [reply packages/edit-site/src/components/editor/index.js:186]: While I don't see the immediate need for that, I'm happy to go ahead and extract it when it can actually be reused somewhere.

File: packages/data/src/redux-store/metadata/selectors.js
Comment (tyxla) [reply packages/data/src/redux-store/metadata/selectors.js:142]: To me, one of the positive sides of metadata selectors is that they are always publicly available. That being said, I wonder if you think there's a particular need for this specific one to be private.

File: packages/edit-site/src/components/editor/index.js
Comment (youknowriad) [reply packages/edit-site/src/components/editor/index.js:186]: it is mostly a suggestion to keep that logic outside the Editor component (timeout...) But I don't feel strongly.

File: packages/data/src/redux-store/metadata/selectors.js
Comment (youknowriad) [reply packages/data/src/redux-store/metadata/selectors.js:142]: Not for me. I think it's fine for this one to be public. I was suggesting it in case we missed confidence.

**Comment by tyxla** [comment]
Thanks, everyone! Looking forward to adding more improvements to the loading experience in the near future as part of https://github.com/WordPress/gutenberg/issues/35503.

File: packages/edit-site/src/components/editor/index.js
Comment (talldan) [reply packages/edit-site/src/components/editor/index.js:178]: Agree with all your points @tyxla 👍 

Great job on getting this merged. 🎉

**Comment by jasmussen** [comment]
Thanks for this one! 🙏

File: packages/edit-site/src/components/editor/index.js
Comment (tyxla) [reply packages/edit-site/src/components/editor/index.js:186]: I think this is a good idea, I'll follow up in another PR.

Comment (tyxla) [reply packages/edit-site/src/components/editor/index.js:186]: Addressed in #50511

File: packages/data/src/redux-store/metadata/selectors.js
Comment (mcsf) [inline packages/data/src/redux-store/metadata/selectors.js:144]: 1. Are the copies necessary? `[ ...Object.values( state ) ].some` -> `Object.values( state ).some`, and likewise for `selectorState`.

 2. After digging around a bit, it seems like `._map` is an internal property of package [`equivalent-key-map`](https://github.com/aduth/equivalent-key-map). Right? My follow-up questions are: 1) is this the right/only way to consume it? 2) If so, shouldn't we at least have a comment explaining how and why we are using a private interface of an external package?

Comment (jsnajdr) [reply packages/data/src/redux-store/metadata/selectors.js:144]: The spread operators would be needed for results of `Map.prototype.values()`, because that's an iterator. But `Object.values()` returns an array, and that doesn't need the extra `[...]`. So, one of the spreads is justified, the other is not.

The `.map.values()` call indeed depends too much on the implementation details, and it seems to me it works only when the selector has exactly one argument? Because then the value is not in `resolution[ 1 ]`, but deeper in the tree. But the only remedy is to contribute a new `values()` method to `equivalent-key-map` itself.

Comment (tyxla) [reply packages/data/src/redux-store/metadata/selectors.js:144]: Thank you for the feedback @mcsf! I've addressed it in #50865. A couple of notes to add:

1. The first copy was indeed unnecessary, but the second one is necessary since the returned iterable iterator object doesn't have a `.some()`, thus we need to convert it to an array. We could use the `EquivalentKeyMap`'s `forEach()` method, however, this would be less optimized since it often will require more iterations than a `.some()`.

2. I've added a comment to explain that usage. As mentioned above, `.some()` is a more optimized usage since it doesn't require iterating through all elements, which is why I prefer it to the `.forEach()` that `EquivalentKeyMap` supports. 

Alternatively, we could open a PR in the `EquivalentKeyMap` repo for supporting `.some()` or a method to just retrieve all elements from the map if you think that will help.

Edit: ha, Jarda beat me to it. Spot on btw!

Comment (jsnajdr) [reply packages/data/src/redux-store/metadata/selectors.js:144]: Note that the standard `Map` doesn't support `.some`. Only `.keys`, `.values`, and `.entries`, and unlike the similar `Object.*` methods they return iterators. Only after converting to array with `[...]` or `Array.from` you can call `.every` or `.some`.

What we'd like to do is to add `.keys`, `.values` and `.entries` to EKM. And I also suspect that their (or, more precisely, Andrew's 🙂) `.size` implementation is wrong, because it returns `._map.size`, but the `._map` is a tree with potentially many nested leaf nodes.

Comment (mcsf) [reply packages/data/src/redux-store/metadata/selectors.js:144]: Thanks for the answers, both of you. Right, I didn't take the time to understand the data structure behind `selectorState`, and indeed `.values` in the context of a map should just be an iterator.

> As mentioned above, .some() is a more optimized usage since it doesn't require iterating through all elements, which is why I prefer it to the .forEach() that EquivalentKeyMap supports.

To be clear, I think `.some` is the best idiom for the situation, and I wouldn't want to replace it with `.forEach`.

I'm fine with just removing the copy around `Object.values` and leaving the other as is. However, just a minor question: personally and intuitively, do you make a difference between `[ ...xs ]` and `Array.from( xs )`? For me, I tend to _infer_ that the _purpose_ of the former is to copy, while that of the latter is to cast (from iterator to array). If I'm the only one with this intuition, then ignore me; if not, then maybe this piece of code becomes 0.1% clearer by using `Array.from( selectorState._map.values ).some`. 😉

Comment (tyxla) [reply packages/data/src/redux-store/metadata/selectors.js:144]: TBH spread and `Array.from` are all the same for me in the current use case (they do have some differences, where `Array.from()` is a bit more powerful and supports more use cases than iterables, but here it doesn't really matter. 

Regardless, I'm happy to make this 0.1% more readable 👍 Updated in the PR.

Comment (jsnajdr) [reply packages/data/src/redux-store/metadata/selectors.js:144]: I usually prefer `Array.from`, because it does the same thing, but is less obscure: it's a regular function call, not a special syntax. Spread operator is more useful when spreading multiple things together.

**Pull request merged**
